iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,603 @@
+        }
+    }
+
+    /// Dispatches messages from lazy queue over to lazy peers.
+    fn on_dispatch_timer(&mut self, io: &mut impl IO<PA>) {
+        for (peer, list) in self.lazy_push_queue.drain() {
+            io.push(OutEvent::SendMessage(peer, Message::IHave(list)));
+        }
+
+        self.dispatch_timer_scheduled = false;
+    }
+
+    /// Send a gossip message.
+    /// Will be pushed in full to eager peers.
+    /// Pushing the message ids to the lazy peers is delayed by a timer.
+    fn do_broadcast(&mut self, data: Bytes, io: &mut impl IO<PA>) {
+        let id = MessageId::from_content(&data);
+        let message = Gossip {
+            id,
+            round: Round(0),
+            content: data,
+        };
+        self.received_messages.insert(id);
+        self.cache.insert(id, message.clone());
+        let me = self.me;
+        self.eager_push(message.clone(), &me, io);
+        self.lazy_push(message, &me, io);
+    }
+
+    fn on_gossip(&mut self, sender: PA, message: Gossip, io: &mut impl IO<PA>) {
+        // if we already received this message: move peer to lazy set
+        // and notify peer about this.
+        if self.received_messages.contains(&message.id) {
+            self.add_lazy(sender);
+            io.push(OutEvent::SendMessage(sender, Message::Prune));
+        // otherwise store the message, emit to application and forward to peers
+        } else {
+            // insert the message in the list of received messages
+            self.received_messages.insert(message.id);
+
+            // increase the round for forwarding the message, and add to cache
+            // to reply to Graft messages later
+            // TODO: use an LRU cache for self.cache
+            // TODO: add callback/event to application to get missing messages that were received before?
+            let message = message.next_round();
+            self.cache.insert(message.id, message.clone());
+
+            // push the message to our peers
+            self.eager_push(message.clone(), &sender, io);
+            self.lazy_push(message.clone(), &sender, io);
+
+            // cleanup places where we track missing messages
+            self.graft_timer_scheduled.remove(&message.id);
+            let previous_ihaves = self.missing_messages.remove(&message.id);
+            // do the optimization step from the paper
+            if let Some(previous_ihaves) = previous_ihaves {
+                self.optimize_tree(&sender, &message, previous_ihaves, io);
+            }
+
+            // emit event to application
+            io.push(OutEvent::EmitEvent(Event::Received(message.content)));
+
+            self.stats.max_last_delivery_hop =
+                self.stats.max_last_delivery_hop.max(message.round.0);
+        }
+    }
+
+    fn optimize_tree(
+        &mut self,
+        sender: &PA,
+        message: &Gossip,
+        previous_ihaves: VecDeque<(PA, Round)>,
+        io: &mut impl IO<PA>,
+    ) {
```

Comment:
my understanding is this should be done with lazy peers, so it would be good to verify the peer is not already eager

Reply:
Fixed in [5a6f2fdd](https://github.com/n0-computer/iroh/pull/1149/commits/5a6f2fddcde3985bc478b64708d39c073a2e5315)